### PR TITLE
Fix Int8(-128) // Int8(-128) 

### LINKED ISSUE #51731: Julia cannot handle a computationally fine division due to implementation details
```
Rational{Int8}(-128) / Rational{Int8}(-128)
```
should give back `Rational{Int8}(1)`, however it produces an error in the implementation
```
ERROR: OverflowError: gcd(-128, -128) overflows
Stacktrace:
 [1] __throw_gcd_overflow(a::Int8, b::Int8)
   @ Base ./intfuncs.jl:57
 [2] gcd(a::Int8, b::Int8)
   @ Base ./intfuncs.jl:54
 [3] divgcd
   @ ./rational.jl:44 [inlined]
 [4] //(x::Rational{Int8}, y::Rational{Int8})
   @ Base ./rational.jl:73
 [5] /(x::Rational{Int8}, y::Rational{Int8})
   @ Base ./rational.jl:360
 [6] top-level scope
   @ REPL[3]:1
```

and this is not inherent to the computation.

Julia installed from the dmg, versioninfo:
```
julia> versioninfo()
Julia Version 1.9.2
Commit e4ee485e909 (2023-07-05 09:39 UTC)
Platform Info:
  OS: macOS (arm64-apple-darwin22.4.0)
  CPU: 12 × Apple M2 Max
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-14.0.6 (ORCJIT, apple-m1)
  Threads: 1 on 8 virtual cores
```

### PR DESCRIPTION
**Pull request by nhz2**
Fixes #51731 and #51730 

Also fixes one of the previously broken tests:
```julia
@test_broken Rational{Int64}(UInt(1), typemin(Int32)) == Int64(1) // Int64(typemin(Int32))
```

This PR ensures the `Rational{T}` constructor with concrete `T` will only throw if the final numerator and denominator cannot be represented by `T`, or are both zero.

If the `T` in `Rational{T}` is not concrete, this PR tries to ensure the numerator and denominator are promoted to the same type.
This means `-1*Rational{Integer}(-1, 0x01) == 1` doesn't throw now. A side effect of this is that `Rational{Integer}(Int8(-1), 0x01)` now throws. 

Also, related to <https://github.com/JuliaLang/julia/pull/25702#issuecomment-359821951>, now `divgcd` doesn't change the types of its inputs and can handle `typemin`, but it still throws a `DivideError` if both inputs are zero.

### CODE REVIEW

File: base/rational.jl [hunk 1]
```diff
@@ -117,7 +129,7 @@ function Rational{T}(x::Rational) where T<:Integer
     unsafe_rational(T, convert(T, x.num), convert(T, x.den))
 end
 function Rational{T}(x::Integer) where T<:Integer
-    unsafe_rational(T, convert(T, x), one(T))
+    unsafe_rational(T, T(x), T(one(x)))
```
[the remark this answers is not in the document]
Comment (nhz2) [reply base/rational.jl:132]: Not sure. Is there an integer type where this would be different?

[the remark this answers is not in the document]
Comment (nhz2) [reply base/rational.jl:132]: The point is to avoid:
```julia
julia> Rational{Integer}(0x01)
0x01//1
```
Where the numerator and denominator can have different types.

[the remark this answers is not in the document]
Comment (nhz2) [reply base/rational.jl:132]: I could change it to `T(oneunit(x))`

[the remark this answers is not in the document]
Comment (giordano) [reply base/rational.jl:132]: `oneunit(T)` is literally the same as `T(one(T))`, but clearer in the intent (hence the suggestion), `T(oneunit(T))` would be a useless duplicate.

[the remark this answers is not in the document]
Comment (nhz2) [reply base/rational.jl:132]: Here though `T` and `typeof(x)` can be different.

[the remark this answers is not in the document]
Comment (giordano) [reply base/rational.jl:132]: Aaah, ok, yes, I missed that bit between all the `Integer`s.

File: base/rational.jl [hunk 2]
```diff
@@ -30,16 +30,27 @@ checked_den(num::Integer, den::Integer) = checked_den(promote(num, den)...)
...
+    end
 end
 
 Rational(n::T, d::T) where {T<:Integer} = Rational{T}(n, d)
 Rational(n::Integer, d::Integer) = Rational(promote(n, d)...)
 Rational(n::Integer) = unsafe_rational(n, one(n))
 
-function divgcd(x::Integer,y::Integer)
-    g = gcd(x,y)
+function divgcd(x::TX, y::TY)::Tuple{TX, TY} where {TX<:Integer, TY<:Integer}
```
Comment (andrewjradcliffe) [inline base/rational.jl:52]: I think that the `divgcd` problem should be addressed in a targeted manner. Specifically, the only case that needs to be handled can be expressed with just 1 additional branch (to be added as a dispatch).

```julia
function divgcd(x::T, y::T) where {T<:Signed}
    a = x == typemin(T)
    b = y == typemin(T)
    if a & b
        one(T), one(T)
    else
        g = gcd(x,y)
        div(x,g), div(y,g)
    end
end
```

Consequently, there is no longer a need to change the Rational constructor. The base problem here has the form `typemin(T) // typemin(T) where T<:Signed`, which can addressed directly without bloating the type constructor. This will make the code easier to maintain and extend in the future.

Comment (nhz2) [reply base/rational.jl:52]: Unfortunately `gcd` will also fail if one input is `typemin` and the other is `zero`.
```julia
julia> gcd(typemin(Int),zero(Int))
ERROR: OverflowError: checked arithmetic: cannot compute |x| for x = -9223372036854775808::Int64
Stacktrace:
 [1] checked_abs
   @ ./checked.jl:128 [inlined]
 [2] gcd(a::Int64, b::Int64)
   @ Base ./intfuncs.jl:52
 [3] top-level scope
   @ REPL[4]:1
```
Also, `typemin` will fail on `BigInt`.
```julia
julia> typemin(BigInt)
ERROR: MethodError: no method matching typemin(::Type{BigInt})
```

Comment (nhz2) [reply base/rational.jl:52]: `gcd` also will fail if one input is `typemin` and signed and the other is unsigned of the same size.
```julia
julia> gcd(Int8(-128), UInt8(0x02))
ERROR: InexactError: check_top_bit(UInt8, -128)
```

Comment (andrewjradcliffe) [reply base/rational.jl:52]: It's true -- I was careless with the `where {T<:Signed}` clause due to `BigInt`. It should instead be `where {T<:BitSigned}`. In fact, I was completely wrong in returning `one(T), one(T)`; I was thinking only of the constructor, not the other uses.

Here is a fixed version which handles all 3 special cases:
```julia
function divgcd(x::T, y::T) where {T<:BitSigned}
    a = x == typemin(T)
    b = y == typemin(T)
    c = iszero(x)
    d = iszero(y)
    if (a & (b | d)) | (c & b) # same as (a & b) | (a & d) | (c & b)
        # `pow` should evaluate to a constant.
        # Why unsigned? because we want to dispatch directly to `ashr_int` and
        # avoid the `ifelse` which would occur if we used `Int`.
        # This is safe as Julia does not have zero-sized types, or,
        # moroever, any type which is a fixed-size integer cannot have
        # a size of zero.
        pow = unsigned(8 * sizeof(T) - 1)
        x >> pow, y >> pow
    else
        g = gcd(x,y)
        div(x,g), div(y,g)
    end
end
```


With respect to `gcd` failing on inputs of mixed types, i.e. where one argument is `Signed` and the other argument `Unsigned`, my thinking is that this is an edge case that deserves to error. The intended semantics when mixing `Signed` and `Unsigned` quickly becomes ambiguous -- it seems better to error than choose an arbitrary behavior.

Comment (andrewjradcliffe) [reply base/rational.jl:52]: If one really wants, then the branch can be removed with predication, but a quick benchmark (and intuition) suggest that it is not worth it here due to the added operations (and because itself will not vectorize). Furthermore, obfuscates the code for the casual reader. Anyway, here's the branchless version:
```julia
function divgcd_branchless(x::T, y::T) where {T<:BitSigned}
    a = x == typemin(T)
    b = y == typemin(T)
    c = iszero(x)
    d = iszero(y)
    s = (a & (b | d)) | (c & b)
    pow = unsigned(8 * sizeof(T) - 1)
    x = (!s * x) + (s * x >> pow)
    y = (!s * y) + (s * y >> pow)
    g = gcd(x,y)
    div(x,g), div(y,g)
end
```

Comment (nhz2) [reply base/rational.jl:52]: Should `Rational{Int}(UInt(1), -2)` also error because it is a mix of `Signed` and `Unsigned`? Or should that only error for specific values and types of unsigned and signed? It would be much simpler if the `Rational{T}` constructor required its arguments to be the same type, but I'm not sure if changing that would be a breaking change.

Comment (nhz2) [reply base/rational.jl:52]: I'm going to try and add a more systematic test of all the edge cases because I think I may be missing some other issues with `Rational{Bool}`.

Comment (andrewjradcliffe) [reply base/rational.jl:52]: Parametric type constructor with explicit parameterization, e.g. `Rational{Int}`, is not as ambiguous as `Rational(UInt(1), -2)` (which does error). There are many ways to break rationals with fixed-size integers.

I would recommend focusing on the original issue PR targeted (as per the its title). Then, one could make a separate PR (or multiple) for the other edge cases you have discovered.

> Also, before this PR Rational{Integer}(Int8(-1), 0x01) didn't throw, but now does.

Breaking existing code is never a good thing; I think it would make this merge a tough sell. There is probably some clever way to handle all the edge cases without breaking the existing interface (whatever it is), but I hypothesize that it will be a few days' work to get it done properly.

Comment (nhz2) [reply base/rational.jl:52]: I added more systematic tests and fixed the `Rational{Bool}(-1,-1)` edge case. I think the ideal interface is:

> The Rational{T} constructor with concrete T will only throw if the final numerator and denominator cannot be represented by T, or are both zero.

The issue with `Rational{Integer}(Int8(-1), 0x01)` is that `Integer` is not concrete, even if this didn't throw before this PR, the resulting number would error often if used, for example:
```julia
julia> a=Rational{Integer}(Int8(-1), 0x01)
-1//0x01

julia> a*Int8(1)
ERROR: InexactError: check_top_bit(UInt8, -1)
```

I think many of the math functions that act on `Rational` implicitly assume the type of the numerator and denominator are the same.

**Review by oscardssmith — APPROVED** [review]
